File: src/classes/grid.cpp
#include "grid.hpp"
#include "simulation.hpp"
#include "particle.hpp"
#include <glm/glm.hpp>
#include <vector>
#include <memory>
#include <iostream>
#include <map>
#include <unordered_map>



Grid::Grid(float cellSize) {
    this->cellSize = cellSize;
}

glm::ivec3 Grid::getCell(glm::vec3 position) {
    return glm::ivec3(position / cellSize); // converting to the largest integer less than or equal to the value
}

void Grid::insert(std::shared_ptr<Sphere> sphere) {
    glm::ivec3 cell = getCell(sphere->position);
    grid[cell].push_back(sphere);
}

void Grid::clear() {
    grid.clear();
}

std::vector<std::shared_ptr<Sphere>> Grid::getNeighbors(std::shared_ptr<Sphere> sphere) {
    std::vector<std::shared_ptr<Sphere>> neighbors;
    glm::ivec3 cell = getCell(sphere->position);
    for (int x = -1; x <= 1; ++x) {
        for (int y = -1; y <= 1; ++y) {
            for (int z = -1; z <= 1; ++z) {
                glm::ivec3 neighbor_cell = cell + glm::ivec3(x, y, z);
                std::vector<std::shared_ptr<Sphere>> temp;
                if (grid.find(neighbor_cell) != grid.end()) {
                    temp = grid[neighbor_cell];
                } else {
                    continue;
                }
                for (auto& s : temp) {
                    if (s != sphere) {
                        neighbors.push_back(s);
                    }
                }
            }
        }
    }
    return neighbors;
}

std::vector<std::shared_ptr<Sphere>> Grid::getNeighbors(glm::ivec3 cell) {
    std::vector<std::shared_ptr<Sphere>> neighbors;
    for (int x = -1; x <= 1; ++x) {
        for (int y = -1; y <= 1; ++y) {
            for (int z = -1; z <= 1; ++z) {
                glm::ivec3 neighbor_cell = cell + glm::ivec3(x, y, z);
                std::vector<std::shared_ptr<Sphere>> temp;
                if (grid.find(neighbor_cell) != grid.end()) {
                    temp = grid[neighbor_cell];
                } else {
                    continue;
                }
                for (auto& s : temp) {
                    neighbors.push_back(s);
                }
            }
        }
    }
    return neighbors;
}

File: rust/src/classes/grid.rs
use nalgebra::Vector3;
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use crate::classes::particles::sphere::Sphere;

#[derive(Hash, Eq, PartialEq, Debug, Clone, Copy)]
pub struct IVec3 {
    x: i32,
    y: i32,
    z: i32,
}

impl From<Vector3<f32>> for IVec3 {
    fn from(vec: Vector3<f32>) -> Self {
        IVec3 {
            x: vec.x.floor() as i32,
            y: vec.y.floor() as i32,
            z: vec.z.floor() as i32,
        }
    }
}

pub struct Grid {
    cell_size: f32,
    pub grid: HashMap<IVec3, Vec<Arc<RwLock<Sphere>>>>,
}

impl Grid {
    pub fn new(cell_size: f32) -> Self {
        Grid {
            cell_size,
            grid: HashMap::new(),
        }
    }

    fn get_cell(&self, position: Vector3<f32>) -> IVec3 {
        IVec3::from(position / self.cell_size)
    }

    pub fn insert(&mut self, sphere: &Arc<RwLock<Sphere>>) {
        let position = sphere.write().unwrap().particle.position;
        let cell = self.get_cell(position);
        self.grid.entry(cell).or_insert_with(Vec::new).push(sphere.clone());
    }

    pub fn clear(&mut self) {
        self.grid.clear();
    }

    // pub fn get_neighbors(&self, sphere: Arc<RwLock<Sphere>>) -> Vec<Arc<RwLock<Sphere>>> {
        // let position = sphere.lock().unwrap().particle.position;
        // let cell = self.get_cell(position);
        // self.get_neighbors_in_cell(cell, Some(sphere))
    // }

    pub fn get_neighbors_in_cell(&self, cell: IVec3) -> Vec<Arc<RwLock<Sphere>>> {
        let mut neighbors = Vec::new();
        for x in -1..=1 {
            for y in -1..=1 {
                for z in -1..=1 {
                    let neighbor_cell = IVec3 {
                        x: cell.x + x,
                        y: cell.y + y,
                        z: cell.z + z,
                    };
                    if let Some(temp) = self.grid.get(&neighbor_cell) {
                        for sphere in temp {
                            neighbors.push(sphere.clone());
                        }
                    }
                }
            }
        }
        neighbors
    }
}